pkgs/development/libraries/mesa/default.nix

```diff
@@ -117,40 +117,26 @@
 }:
 
 let
-  rustDeps = [
-    {
-      pname = "paste";
-      version = "1.0.14";
-      hash = "sha256-+J1h7New5MEclUBvwDQtTYJCHKKqAEOeQkuKy+g0vEc=";
-    }
-    {
-      pname = "proc-macro2";
-      version = "1.0.86";
-      hash = "sha256-9fYAlWRGVIwPp8OKX7Id84Kjt8OoN2cANJ/D9ZOUUZE=";
-    }
-    {
-      pname = "quote";
-      version = "1.0.33";
-      hash = "sha256-VWRCZJO0/DJbNu0/V9TLaqlwMot65YjInWT9VWg57DY=";
-    }
-    {
-      pname = "syn";
-      version = "2.0.68";
-      hash = "sha256-nGLBbxR0DFBpsXMngXdegTm/o13FBS6QsM7TwxHXbgQ=";
-    }
-    {
-      pname = "unicode-ident";
-      version = "1.0.12";
-      hash = "sha256-KX8NqYYw6+rGsoR9mdZx8eT1HIPEUUyxErdk2H/Rlj8=";
-    }
-  ];
+  rustDeps = lib.importJSON ./wraps.json;
+
+  handleRustDep = dep: ''
+    (
+      echo "Preparing Rust dependency wrap ${dep.pname}-${dep.version}..."
+
+      wrapDir=subprojects/packagefiles/${dep.pname}
+      outDir=subprojects/${dep.pname}-${dep.version}
 
-  copyRustDep = dep: ''
-    cp -R --no-preserve=mode,ownership ${fetchCrate dep} subprojects/${dep.pname}-${dep.version}
-    cp -R subprojects/packagefiles/${dep.pname}/* subprojects/${dep.pname}-${dep.version}/
+      mkdir -p $outDir
+      tar -C $outDir -xf ${fetchCrate (dep // { unpack = false; })} --strip-components=1
+      cp $wrapDir/meson.build $outDir
+
+      for patch in $wrapDir/*.patch; do
+        patch -d $outDir -i $(readlink -f $patch) -p1
+      done
+    )
```

Comment:
We're doing way more work here than we need to. Meson supports a `MESON_PACKAGE_CACHE` environment variable, pointing to a directory of tarballs, which can be symlinks. See https://spectrum-os.org/git/spectrum/tree/tools/default.nix?id=cf7c311bd1ec6e7283e0a65d65b2e7ea587bb49e#n83 (which is also worse than it needs to be because it predates the existence of that environment variable).

Reply:
https://github.com/mesonbuild/meson/blob/37f7572ce15494654128184777b16ece38308b9a/mesonbuild/wrap/wrap.py#L515-L521

OK I guess that's fairly frozen?

Reply:
OK that does make it significantly less cursed.